driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java

```diff
@@ -120,31 +124,33 @@ public void run() {
 
         Executor initExecutor = manager.cluster.manager.configuration.getPoolingOptions().getInitializationExecutor();
 
-        ListenableFuture<List<Void>> allConnectionsFuture = Futures.allAsList(connectionFutures);
+        ListenableFuture<List<Void>> allConnectionsFuture = Futures.successfulAsList(connectionFutures);
 
         final SettableFuture<Void> initFuture = SettableFuture.create();
-        Futures.addCallback(allConnectionsFuture, new FutureCallback<List<Void>>() {
+        Futures.addCallback(allConnectionsFuture, new MoreFutures.SuccessCallback<List<Void>>() {
             @Override
             public void onSuccess(List<Void> l) {
+                // Some of the connections might have failed, keep only the successful ones
+                ListIterator<Connection> it = connections.listIterator();
```

Comment:
It looks like this works in all cases on initialization except when we get an UnsupportedProtocolVersion/ClusterNameMismatchException because we don't defunct the connection in these cases (and thus don't call signalConnectionFailure in those cases, so the host remains up:

```java
        // Fallback on initializeTransportFuture so we can properly propagate specific exceptions.
        ListenableFuture<Void> initFuture = Futures.withFallback(initializeTransportFuture, new FutureFallback<Void>() {
            @Override
            public ListenableFuture<Void> create(Throwable t) throws Exception {
                SettableFuture<Void> future = SettableFuture.create();
                // Make sure the connection gets properly closed.
                if (t instanceof ClusterNameMismatchException || t instanceof UnsupportedProtocolVersionException) {
                    // These exceptions cause the node to be ignored, so just propagate
                    closeAsync().force();
                    future.setException(t);
                } else {
                    // Defunct to ensure that the error will be signaled (marking the host down)
                    Exception e = (t instanceof ConnectionException || t instanceof DriverException || t instanceof InterruptedException)
                        ? (Exception)t
                        : new ConnectionException(Connection.this.address,
                        String.format("Unexpected error during transport initialization (%s)", t),
                        t);
                    future.setException(defunct(e));
                }
                return future;
            }
        }, initExecutor);
```

If I add signalConnectionFailure in the if branch for ClusterNameMismatch/UnsupportedProtocolVersion the host is now marked down, but not sure if that is the appropriate fix.